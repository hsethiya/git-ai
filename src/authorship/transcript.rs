use serde::{Deserialize, Serialize};

/// Represents a single message in an AI transcript
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Message {
    User {
        text: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        timestamp: Option<String>,
    },
    Assistant {
        text: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        timestamp: Option<String>,
    },
    ToolUse {
        name: String,
        input: serde_json::Value,
        #[serde(skip_serializing_if = "Option::is_none")]
        timestamp: Option<String>,
    },
}

impl Message {
    /// Create a user message
    pub fn user(text: String, timestamp: Option<String>) -> Self {
        Message::User { text, timestamp }
    }

    /// Create an assistant message
    pub fn assistant(text: String, timestamp: Option<String>) -> Self {
        Message::Assistant { text, timestamp }
    }

    /// Create a tool use message
    pub fn tool_use(name: String, input: serde_json::Value) -> Self {
        Message::ToolUse {
            name,
            input,
            timestamp: None,
        }
    }

    /// Get the text content if this is a user or assistant message
    #[allow(dead_code)]
    pub fn text(&self) -> Option<&String> {
        match self {
            Message::User { text, .. } | Message::Assistant { text, .. } => Some(text),
            Message::ToolUse { .. } => None,
        }
    }

    /// Check if this is a tool use message
    #[allow(dead_code)]
    pub fn is_tool_use(&self) -> bool {
        matches!(self, Message::ToolUse { .. })
    }
}

/// Represents a complete AI transcript (collection of messages)
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AiTranscript {
    pub messages: Vec<Message>,
}

impl AiTranscript {
    /// Create a new empty transcript
    pub fn new() -> Self {
        Self {
            messages: Vec::new(),
        }
    }

    /// Add a message to the transcript
    pub fn add_message(&mut self, message: Message) {
        self.messages.push(message);
    }

    /// Get all messages
    pub fn messages(&self) -> &[Message] {
        &self.messages
    }

    /// Filter out tool use messages
    #[allow(dead_code)]
    pub fn without_tool_use(&self) -> Self {
        let filtered_messages: Vec<Message> = self
            .messages
            .iter()
            .filter(|msg| !msg.is_tool_use())
            .cloned()
            .collect();

        Self {
            messages: filtered_messages,
        }
    }

    /// Parse a Claude Code JSONL file into a transcript and extract model info
    pub fn from_claude_code_jsonl_with_model(
        jsonl_content: &str,
    ) -> Result<(Self, Option<String>), serde_json::Error> {
        let mut transcript = AiTranscript::new();
        let mut model = None;

        for line in jsonl_content.lines() {
            if !line.trim().is_empty() {
                // Parse the raw JSONL entry
                let raw_entry: serde_json::Value = serde_json::from_str(line)?;
                let timestamp = raw_entry["timestamp"].as_str().map(|s| s.to_string());

                // Extract model from assistant messages if we haven't found it yet
                if model.is_none() && raw_entry["type"].as_str() == Some("assistant") {
                    if let Some(model_str) = raw_entry["message"]["model"].as_str() {
                        model = Some(model_str.to_string());
                    }
                }

                // Extract messages based on the type
                match raw_entry["type"].as_str() {
                    Some("user") => {
                        // Handle user messages
                        if let Some(content) = raw_entry["message"]["content"].as_str() {
                            if !content.trim().is_empty() {
                                transcript.add_message(Message::User {
                                    text: content.to_string(),
                                    timestamp: timestamp.clone(),
                                });
                            }
                        } else if let Some(content_array) =
                            raw_entry["message"]["content"].as_array()
                        {
                            // Handle user messages with content array (like tool results)
                            for item in content_array {
                                if let Some(text) = item["content"].as_str() {
                                    if !text.trim().is_empty() {
                                        transcript.add_message(Message::User {
                                            text: text.to_string(),
                                            timestamp: timestamp.clone(),
                                        });
                                    }
                                }
                            }
                        }
                    }
                    Some("assistant") => {
                        // Handle assistant messages
                        if let Some(content_array) = raw_entry["message"]["content"].as_array() {
                            for item in content_array {
                                match item["type"].as_str() {
                                    Some("text") => {
                                        if let Some(text) = item["text"].as_str() {
                                            if !text.trim().is_empty() {
                                                transcript.add_message(Message::Assistant {
                                                    text: text.to_string(),
                                                    timestamp: timestamp.clone(),
                                                });
                                            }
                                        }
                                    }
                                    Some("tool_use") => {
                                        if let (Some(name), Some(_input)) =
                                            (item["name"].as_str(), item["input"].as_object())
                                        {
                                            transcript.add_message(Message::ToolUse {
                                                name: name.to_string(),
                                                input: item["input"].clone(),
                                                timestamp: timestamp.clone(),
                                            });
                                        }
                                    }
                                    _ => continue, // Skip unknown content types
                                }
                            }
                        }
                    }
                    _ => continue, // Skip unknown message types
                }
            }
        }

        Ok((transcript, model))
    }
}

impl Default for AiTranscript {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    #[test]
    fn test_parse_example_claude_code_jsonl_with_model() {
        let jsonl_content = fs::read_to_string(
            "src/commands/checkpoint_agent/claude_code/test_data/example-claude-code.jsonl",
        )
        .expect("Failed to read example JSONL file");

        let (transcript, model) = AiTranscript::from_claude_code_jsonl_with_model(&jsonl_content)
            .expect("Failed to parse JSONL");

        // Verify we parsed some messages
        assert!(!transcript.messages().is_empty());

        // Verify we extracted the model
        assert!(model.is_some());
        let model_name = model.unwrap();
        println!("Extracted model: {}", model_name);

        // Based on the example file, we should get claude-sonnet-4-20250514
        assert_eq!(model_name, "claude-sonnet-4-20250514");

        // Print the parsed transcript for inspection
        println!("Parsed {} messages:", transcript.messages().len());
        for (i, message) in transcript.messages().iter().enumerate() {
            match message {
                Message::User { text, .. } => println!("{}: User: {}", i, text),
                Message::Assistant { text, .. } => println!("{}: Assistant: {}", i, text),
                Message::ToolUse { name, input, .. } => {
                    println!("{}: ToolUse: {} with input: {:?}", i, name, input)
                }
            }
        }
    }
}
