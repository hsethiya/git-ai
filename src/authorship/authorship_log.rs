use crate::authorship::transcript::Message;
use crate::authorship::working_log::AgentId;
use serde::{Deserialize, Serialize};
use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Author {
    pub username: String,
    pub email: String,
}

/// Represents either a single line or a range of lines
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub enum LineRange {
    Single(u32),
    Range(u32, u32), // start, end (inclusive)
}

impl LineRange {
    pub fn contains(&self, line: u32) -> bool {
        match self {
            LineRange::Single(l) => *l == line,
            LineRange::Range(start, end) => line >= *start && line <= *end,
        }
    }

    #[allow(dead_code)]
    pub fn overlaps(&self, other: &LineRange) -> bool {
        match (self, other) {
            (LineRange::Single(l1), LineRange::Single(l2)) => l1 == l2,
            (LineRange::Single(l), LineRange::Range(start, end)) => *l >= *start && *l <= *end,
            (LineRange::Range(start, end), LineRange::Single(l)) => *l >= *start && *l <= *end,
            (LineRange::Range(start1, end1), LineRange::Range(start2, end2)) => {
                start1 <= end2 && start2 <= end1
            }
        }
    }

    /// Remove a line or range from this range, returning the remaining parts
    #[allow(dead_code)]
    pub fn remove(&self, to_remove: &LineRange) -> Vec<LineRange> {
        match (self, to_remove) {
            (LineRange::Single(l), LineRange::Single(r)) => {
                if l == r {
                    vec![]
                } else {
                    vec![self.clone()]
                }
            }
            (LineRange::Single(l), LineRange::Range(start, end)) => {
                if *l >= *start && *l <= *end {
                    vec![]
                } else {
                    vec![self.clone()]
                }
            }
            (LineRange::Range(start, end), LineRange::Single(r)) => {
                if *r < *start || *r > *end {
                    vec![self.clone()]
                } else if *r == *start && *r == *end {
                    vec![]
                } else if *r == *start {
                    vec![LineRange::Range(*start + 1, *end)]
                } else if *r == *end {
                    vec![LineRange::Range(*start, *end - 1)]
                } else {
                    vec![
                        LineRange::Range(*start, *r - 1),
                        LineRange::Range(*r + 1, *end),
                    ]
                }
            }
            (LineRange::Range(start1, end1), LineRange::Range(start2, end2)) => {
                if *start2 > *end1 || *end2 < *start1 {
                    // No overlap
                    vec![self.clone()]
                } else {
                    let mut result = Vec::new();
                    // Left part
                    if *start1 < *start2 {
                        result.push(LineRange::Range(*start1, *start2 - 1));
                    }
                    // Right part
                    if *end1 > *end2 {
                        result.push(LineRange::Range(*end2 + 1, *end1));
                    }
                    result
                }
            }
        }
    }

    /// Convert a sorted list of line numbers into compressed ranges
    pub fn compress_lines(lines: &[u32]) -> Vec<LineRange> {
        if lines.is_empty() {
            return vec![];
        }

        let mut ranges = Vec::new();
        let mut current_start = lines[0];
        let mut current_end = lines[0];

        for &line in &lines[1..] {
            if line == current_end + 1 {
                current_end = line;
            } else {
                // End current range and start new one
                if current_start == current_end {
                    ranges.push(LineRange::Single(current_start));
                } else {
                    ranges.push(LineRange::Range(current_start, current_end));
                }
                current_start = line;
                current_end = line;
            }
        }

        // Add the last range
        if current_start == current_end {
            ranges.push(LineRange::Single(current_start));
        } else {
            ranges.push(LineRange::Range(current_start, current_end));
        }

        ranges
    }

    #[allow(dead_code)]
    pub fn expand(&self) -> Vec<u32> {
        match self {
            LineRange::Single(l) => vec![*l],
            LineRange::Range(start, end) => (*start..=*end).collect(),
        }
    }

    /// Shift line numbers by a given offset
    /// - For insertions: offset is positive (shift lines down/forward)
    /// - For deletions: offset is negative (shift lines up/backward)
    /// - insertion_point: the line number where the change occurred
    #[allow(dead_code)]
    pub fn shift(&self, insertion_point: u32, offset: i32) -> Option<LineRange> {
        match self {
            LineRange::Single(l) => {
                if *l >= insertion_point {
                    let new_line = (*l as i32 + offset) as u32;
                    Some(LineRange::Single(new_line))
                } else {
                    Some(LineRange::Single(*l))
                }
            }
            LineRange::Range(start, end) => {
                let new_start = if *start >= insertion_point {
                    (*start as i32 + offset) as u32
                } else {
                    *start
                };
                let new_end = if *end >= insertion_point {
                    (*end as i32 + offset) as u32
                } else {
                    *end
                };

                // Ensure the range is still valid
                if new_start <= new_end {
                    if new_start == new_end {
                        Some(LineRange::Single(new_start))
                    } else {
                        Some(LineRange::Range(new_start, new_end))
                    }
                } else {
                    None
                }
            }
        }
    }
}

impl fmt::Display for LineRange {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            LineRange::Single(l) => write!(f, "{}", l),
            LineRange::Range(start, end) => write!(f, "[{}, {}]", start, end),
        }
    }
}

/// Prompt session details stored in the top-level prompts map keyed by short hash (agent_id + tool)
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PromptRecord {
    pub agent_id: AgentId,
    pub human_author: Option<String>,
    pub messages: Vec<Message>,
    #[serde(default)]
    pub total_additions: u32,
    #[serde(default)]
    pub total_deletions: u32,
    #[serde(default)]
    pub accepted_lines: u32,
    #[serde(default)]
    pub overriden_lines: u32,
}
