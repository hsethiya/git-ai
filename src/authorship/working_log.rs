use crate::authorship::attribution_tracker::{Attribution, LineAttribution};
use crate::authorship::transcript::AiTranscript;
use serde::{Deserialize, Serialize};
use std::fmt;
use std::time::{SystemTime, UNIX_EPOCH};

pub const CHECKPOINT_API_VERSION: &str = "checkpoint/1.0.0";

/// Represents a working log entry for a specific file
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkingLogEntry {
    /// The file path relative to the repository root
    pub file: String,
    /// SHA256 hash of the file content at this checkpoint
    #[serde(default)]
    pub blob_sha: String,
    #[serde(default)]
    pub attributions: Vec<Attribution>,
    #[serde(default)]
    pub line_attributions: Vec<LineAttribution>,
}

impl WorkingLogEntry {
    /// Create a new working log entry
    pub fn new(
        file: String,
        blob_sha: String,
        attributions: Vec<Attribution>,
        line_attributions: Vec<LineAttribution>,
    ) -> Self {
        Self {
            file,
            blob_sha,
            attributions,
            line_attributions,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AgentId {
    pub tool: String, // e.g., "cursor", "windsurf"
    pub id: String,   // id in their domain
    pub model: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CheckpointKind {
    Human,
    AiAgent,
    AiTab,
}

impl fmt::Display for CheckpointKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.to_str())
    }
}

impl CheckpointKind {
    pub fn from_str(s: &str) -> Self {
        match s {
            "human" => CheckpointKind::Human,
            "ai_agent" => CheckpointKind::AiAgent,
            "ai_tab" => CheckpointKind::AiTab,
            _ => panic!("Invalid checkpoint kind: {}", s),
        }
    }

    pub fn to_str(&self) -> String {
        match self {
            CheckpointKind::Human => "human".to_string(),
            CheckpointKind::AiAgent => "ai_agent".to_string(),
            CheckpointKind::AiTab => "ai_tab".to_string(),
        }
    }

    /// Default value to prevent crashes on old versions
    pub fn serde_default() -> Self {
        CheckpointKind::Human
    }
}

/// Line-level statistics tracked per checkpoint kind
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(default)]
pub struct CheckpointLineStats {
    pub human_additions: u32,
    pub human_deletions: u32,
    pub ai_agent_additions: u32,
    pub ai_agent_deletions: u32,
    pub ai_tab_additions: u32,
    pub ai_tab_deletions: u32,
    pub overrides: u32,
}

impl CheckpointLineStats {
    pub fn additions_for_kind(&self, kind: CheckpointKind) -> u32 {
        match kind {
            CheckpointKind::Human => self.human_additions,
            CheckpointKind::AiAgent => self.ai_agent_additions,
            CheckpointKind::AiTab => self.ai_tab_additions,
        }
    }

    pub fn deletions_for_kind(&self, kind: CheckpointKind) -> u32 {
        match kind {
            CheckpointKind::Human => self.human_deletions,
            CheckpointKind::AiAgent => self.ai_agent_deletions,
            CheckpointKind::AiTab => self.ai_tab_deletions,
        }
    }

    /// Total AI additions (for authorship log - collapses ai_agent and ai_tab)
    pub fn total_ai_additions(&self) -> u32 {
        self.ai_agent_additions + self.ai_tab_additions
    }

    /// Total AI deletions (for authorship log - collapses ai_agent and ai_tab)
    pub fn total_ai_deletions(&self) -> u32 {
        self.ai_agent_deletions + self.ai_tab_deletions
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Checkpoint {
    #[serde(default = "CheckpointKind::serde_default")]
    pub kind: CheckpointKind,
    pub diff: String,
    pub author: String,
    pub entries: Vec<WorkingLogEntry>,
    pub timestamp: u64,
    pub transcript: Option<AiTranscript>,
    pub agent_id: Option<AgentId>,
    #[serde(default)]
    pub line_stats: CheckpointLineStats,
    #[serde(default)]
    pub api_version: String,
}

impl Checkpoint {
    pub fn new(
        kind: CheckpointKind,
        diff: String,
        author: String,
        entries: Vec<WorkingLogEntry>,
    ) -> Self {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs();

        Self {
            kind,
            diff,
            author,
            entries,
            timestamp,
            transcript: None,
            agent_id: None,
            line_stats: CheckpointLineStats::default(),
            api_version: CHECKPOINT_API_VERSION.to_string(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::authorship::transcript::Message;

    #[test]
    fn test_checkpoint_serialization() {
        let entry = WorkingLogEntry::new(
            "src/xyz.rs".to_string(),
            "abc123def456".to_string(),
            Vec::new(),
            Vec::new(),
        );
        let checkpoint = Checkpoint::new(
            CheckpointKind::AiAgent,
            "".to_string(),
            "claude".to_string(),
            vec![entry],
        );

        // Verify timestamp is set (should be recent)
        let current_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs();
        assert!(checkpoint.timestamp > 0);
        assert!(checkpoint.timestamp <= current_time);
        assert!(checkpoint.transcript.is_none());
        assert!(checkpoint.agent_id.is_none());

        let json = serde_json::to_string_pretty(&checkpoint).unwrap();
        let deserialized: Checkpoint = serde_json::from_str(&json).unwrap();
        assert_eq!(deserialized.diff, "");
        assert_eq!(deserialized.entries.len(), 1);
        assert_eq!(deserialized.entries[0].file, "src/xyz.rs");
        assert_eq!(deserialized.entries[0].blob_sha, "abc123def456");
        assert_eq!(deserialized.timestamp, checkpoint.timestamp);
        assert!(deserialized.transcript.is_none());
        assert!(deserialized.agent_id.is_none());
    }

    #[test]
    fn test_log_array_serialization() {
        let entry1 = WorkingLogEntry::new(
            "src/xyz.rs".to_string(),
            "sha1".to_string(),
            Vec::new(),
            Vec::new(),
        );
        let checkpoint1 = Checkpoint::new(
            CheckpointKind::AiAgent,
            "".to_string(),
            "claude".to_string(),
            vec![entry1],
        );

        let entry2 = WorkingLogEntry::new(
            "src/xyz.rs".to_string(),
            "sha2".to_string(),
            Vec::new(),
            Vec::new(),
        );
        let checkpoint2 = Checkpoint::new(
            CheckpointKind::AiAgent,
            "/refs/ai/working/xyz.patch".to_string(),
            "user".to_string(),
            vec![entry2],
        );

        // Verify timestamps are set and checkpoint2 is newer than checkpoint1
        assert!(checkpoint1.timestamp > 0);
        assert!(checkpoint2.timestamp > 0);
        assert!(checkpoint2.timestamp >= checkpoint1.timestamp);

        let log = vec![checkpoint1, checkpoint2];
        let json = serde_json::to_string_pretty(&log).unwrap();
        // println!("Working log array JSON:\n{}", json);
        let deserialized: Vec<Checkpoint> = serde_json::from_str(&json).unwrap();
        assert_eq!(deserialized.len(), 2);
        assert_eq!(deserialized[0].diff, "");
        assert_eq!(deserialized[1].diff, "/refs/ai/working/xyz.patch");
        assert_eq!(deserialized[1].author, "user");
    }

    #[test]
    fn test_checkpoint_with_transcript() {
        let entry = WorkingLogEntry::new(
            "src/xyz.rs".to_string(),
            "test_sha".to_string(),
            Vec::new(),
            Vec::new(),
        );

        let user_message = Message::user(
            "Please add error handling to this function".to_string(),
            None,
        );
        let assistant_message =
            Message::assistant("I'll add error handling to the function.".to_string(), None);

        let mut transcript = AiTranscript::new();
        transcript.add_message(user_message);
        transcript.add_message(assistant_message);

        let agent_id = AgentId {
            tool: "cursor".to_string(),
            model: "gpt-4o".to_string(),
            id: "session-abc123".to_string(),
        };

        let mut checkpoint = Checkpoint::new(
            CheckpointKind::AiAgent,
            "".to_string(),
            "claude".to_string(),
            vec![entry],
        );
        checkpoint.transcript = Some(transcript);
        checkpoint.agent_id = Some(agent_id);

        assert!(checkpoint.transcript.is_some());
        assert!(checkpoint.agent_id.is_some());

        let transcript_data = checkpoint.transcript.as_ref().unwrap();
        assert_eq!(transcript_data.messages().len(), 2);

        // Check first message (user)
        match &transcript_data.messages()[0] {
            Message::User { text, .. } => {
                assert_eq!(text, "Please add error handling to this function");
            }
            _ => panic!("Expected user message"),
        }

        // Check second message (assistant)
        match &transcript_data.messages()[1] {
            Message::Assistant { text, .. } => {
                assert_eq!(text, "I'll add error handling to the function.");
            }
            _ => panic!("Expected assistant message"),
        }

        let agent_data = checkpoint.agent_id.as_ref().unwrap();
        assert_eq!(agent_data.tool, "cursor");
        assert_eq!(agent_data.id, "session-abc123");

        let json = serde_json::to_string_pretty(&checkpoint).unwrap();
        let deserialized: Checkpoint = serde_json::from_str(&json).unwrap();
        assert!(deserialized.transcript.is_some());
        assert!(deserialized.agent_id.is_some());

        let deserialized_transcript = deserialized.transcript.as_ref().unwrap();
        assert_eq!(deserialized_transcript.messages().len(), 2);

        let deserialized_agent = deserialized.agent_id.as_ref().unwrap();
        assert_eq!(deserialized_agent.tool, "cursor");
        assert_eq!(deserialized_agent.id, "session-abc123");
    }
}
