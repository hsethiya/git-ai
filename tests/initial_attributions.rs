mod repos;

use git_ai::authorship::attribution_tracker::LineAttribution;
use git_ai::authorship::authorship_log::PromptRecord;
use git_ai::authorship::transcript::Message;
use git_ai::authorship::working_log::AgentId;
use insta::assert_debug_snapshot;
use regex::Regex;
use repos::test_repo::TestRepo;
use std::collections::HashMap;

/// Normalize blame output for snapshot testing by replacing non-deterministic
/// elements (commit SHAs and timestamps) with placeholders
fn normalize_blame_output(blame_output: &str) -> String {
    // Replace commit SHAs (40 hex chars) with placeholder
    let re_sha = Regex::new(r"[0-9a-f]{40}|[0-9a-f]{7,}").unwrap();
    let result = re_sha.replace_all(blame_output, "COMMIT_SHA");

    // Replace timestamps (e.g., "2025-10-27 11:29:32 -0400") with placeholder
    let re_timestamp = Regex::new(r"\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2} [\+\-]\d{4}").unwrap();
    let result = re_timestamp.replace_all(&result, "TIMESTAMP");

    result.to_string()
}

#[test]
fn test_initial_only_no_blame_data() {
    // Test that INITIAL attributions work when there's no blame data (new file case)
    let repo = TestRepo::new();

    // Create initial commit to have a HEAD
    let mut readme = repo.filename("README.md");
    readme.set_contents(vec!["# Test Repo".to_string()]);
    repo.stage_all_and_commit("initial commit")
        .expect("commit should succeed");

    // Get the working log for current HEAD
    let working_log = repo.current_working_logs();

    // IMPORTANT: Write INITIAL file BEFORE making any file changes
    let mut initial_attributions = HashMap::new();
    let mut line_attrs = Vec::new();
    line_attrs.push(LineAttribution {
        start_line: 1,
        end_line: 3,
        author_id: "initial-ai-123".to_string(),
        overridden: false,
    });
    initial_attributions.insert("newfile.txt".to_string(), line_attrs);

    let mut prompts = HashMap::new();
    prompts.insert(
        "initial-ai-123".to_string(),
        PromptRecord {
            agent_id: AgentId {
                tool: "test-tool".to_string(),
                id: "session-123".to_string(),
                model: "test-model".to_string(),
            },
            human_author: None,
            messages: vec![Message::assistant("Initial attribution".to_string(), None)],
            total_additions: 0,
            total_deletions: 0,
            accepted_lines: 0,
            overriden_lines: 0,
        },
    );

    working_log
        .write_initial_attributions(initial_attributions, prompts)
        .expect("write initial attributions should succeed");

    // NOW create the new file in working directory (this will trigger checkpoint reading)
    let file_content = "line 1 from INITIAL\nline 2 from INITIAL\nline 3 from INITIAL\n";
    std::fs::write(repo.path().join("newfile.txt"), file_content)
        .expect("write file should succeed");

    // Run checkpoint - should use INITIAL attributions since there's no blame data
    repo.git_ai(&["checkpoint"])
        .expect("checkpoint should succeed");

    // Commit and verify
    let commit = repo
        .stage_all_and_commit("add newfile")
        .expect("commit should succeed");

    eprintln!(
        "Authorship log prompts: {:?}",
        commit
            .authorship_log
            .metadata
            .prompts
            .keys()
            .collect::<Vec<_>>()
    );
    eprintln!(
        "Authorship log attestations: {:?}",
        commit
            .authorship_log
            .attestations
            .iter()
            .map(|a| &a.file_path)
            .collect::<Vec<_>>()
    );

    let blame_output = repo
        .git_ai(&["blame", "newfile.txt"])
        .expect("blame should succeed");

    let normalized = normalize_blame_output(&blame_output);
    assert_debug_snapshot!(normalized);
}

#[test]
fn test_initial_wins_overlaps() {
    // Test that INITIAL attributions seed the initial state
    let repo = TestRepo::new();

    // Create initial commit to have a HEAD
    let mut readme = repo.filename("README.md");
    readme.set_contents(vec!["# Test Repo".to_string()]);
    repo.stage_all_and_commit("initial commit")
        .expect("commit should succeed");

    // Get the working log for current HEAD
    let working_log = repo.current_working_logs();

    // IMPORTANT: Write INITIAL file BEFORE creating the file
    let mut initial_attributions = HashMap::new();
    let mut line_attrs = Vec::new();
    line_attrs.push(LineAttribution {
        start_line: 1,
        end_line: 2,
        author_id: "initial-override-456".to_string(),
        overridden: false,
    });
    initial_attributions.insert("example.txt".to_string(), line_attrs);

    let mut prompts = HashMap::new();
    prompts.insert(
        "initial-override-456".to_string(),
        PromptRecord {
            agent_id: AgentId {
                tool: "override-tool".to_string(),
                id: "override-session".to_string(),
                model: "override-model".to_string(),
            },
            human_author: None,
            messages: vec![Message::assistant("Override attribution".to_string(), None)],
            total_additions: 0,
            total_deletions: 0,
            accepted_lines: 0,
            overriden_lines: 0,
        },
    );

    working_log
        .write_initial_attributions(initial_attributions, prompts)
        .expect("write initial attributions should succeed");

    // NOW create the file - INITIAL will seed the checkpoint
    let file_content = "line 1\nline 2\nline 3 modified\n";
    std::fs::write(repo.path().join("example.txt"), file_content)
        .expect("write file should succeed");

    // Run checkpoint
    repo.git_ai(&["checkpoint"])
        .expect("checkpoint should succeed");

    // Commit
    repo.stage_all_and_commit("add example.txt")
        .expect("commit should succeed");

    let blame_output = repo
        .git_ai(&["blame", "example.txt"])
        .expect("blame should succeed");

    let normalized = normalize_blame_output(&blame_output);
    assert_debug_snapshot!(normalized);
}

#[test]
fn test_initial_and_blame_merge() {
    // Test that INITIAL covers some lines and blame fills in the gaps
    let repo = TestRepo::new();

    // Create initial commit to have a HEAD
    let mut readme = repo.filename("README.md");
    readme.set_contents(vec!["# Test Repo".to_string()]);
    repo.stage_all_and_commit("initial commit")
        .expect("commit should succeed");

    // Get the working log for current HEAD
    let working_log = repo.current_working_logs();

    // IMPORTANT: Write INITIAL file BEFORE creating the file
    // INITIAL covers lines 1-3 and 5, blame will be used for lines 4, 6, 7
    let mut initial_attributions = HashMap::new();
    let mut line_attrs = Vec::new();
    line_attrs.push(LineAttribution {
        start_line: 1,
        end_line: 3,
        author_id: "initial-123".to_string(),
        overridden: false,
    });
    line_attrs.push(LineAttribution {
        start_line: 5,
        end_line: 5,
        author_id: "initial-456".to_string(),
        overridden: false,
    });
    initial_attributions.insert("example.txt".to_string(), line_attrs);

    let mut prompts = HashMap::new();
    prompts.insert(
        "initial-123".to_string(),
        PromptRecord {
            agent_id: AgentId {
                tool: "tool1".to_string(),
                id: "session1".to_string(),
                model: "model1".to_string(),
            },
            human_author: None,
            messages: vec![Message::assistant("Attribution 123".to_string(), None)],
            total_additions: 0,
            total_deletions: 0,
            accepted_lines: 0,
            overriden_lines: 0,
        },
    );
    prompts.insert(
        "initial-456".to_string(),
        PromptRecord {
            agent_id: AgentId {
                tool: "tool2".to_string(),
                id: "session2".to_string(),
                model: "model2".to_string(),
            },
            human_author: None,
            messages: vec![Message::assistant("Attribution 456".to_string(), None)],
            total_additions: 0,
            total_deletions: 0,
            accepted_lines: 0,
            overriden_lines: 0,
        },
    );

    working_log
        .write_initial_attributions(initial_attributions, prompts)
        .expect("write initial attributions should succeed");

    // NOW create the file - INITIAL will seed lines 1-3, 5; blame will be used for 4, 6, 7
    // Write directly to filesystem for direct control
    let file_content = "line 1\nline 2\nline 3\nline 4\nline 5\nline 6\nline 7\n";
    std::fs::write(repo.path().join("example.txt"), file_content)
        .expect("write file should succeed");

    // Run checkpoint with mock_ai so lines 4, 6, 7 get mock attribution
    repo.git_ai(&["checkpoint", "mock_ai"])
        .expect("checkpoint should succeed");

    // Commit
    repo.stage_all_and_commit("add example.txt")
        .expect("commit should succeed");

    let blame_output = repo
        .git_ai(&["blame", "example.txt"])
        .expect("blame should succeed");

    let normalized = normalize_blame_output(&blame_output);
    assert_debug_snapshot!(normalized);
}

#[test]
fn test_partial_file_coverage() {
    // Test that INITIAL has data for fileA but not fileB
    let repo = TestRepo::new();

    // Create initial commit to have a HEAD
    let mut readme = repo.filename("README.md");
    readme.set_contents(vec!["# Test Repo".to_string()]);
    repo.stage_all_and_commit("initial commit")
        .expect("commit should succeed");

    // Get the working log for current HEAD
    let working_log = repo.current_working_logs();

    // IMPORTANT: Write INITIAL file AFTER creating initial commit
    let mut initial_attributions = HashMap::new();
    let mut line_attrs = Vec::new();
    line_attrs.push(LineAttribution {
        start_line: 1,
        end_line: 2,
        author_id: "initial-fileA".to_string(),
        overridden: false,
    });
    initial_attributions.insert("fileA.txt".to_string(), line_attrs);
    // Note: fileB.txt is not in INITIAL

    let mut prompts = HashMap::new();
    prompts.insert(
        "initial-fileA".to_string(),
        PromptRecord {
            agent_id: AgentId {
                tool: "toolA".to_string(),
                id: "sessionA".to_string(),
                model: "modelA".to_string(),
            },
            human_author: None,
            messages: vec![Message::assistant("FileA attribution".to_string(), None)],
            total_additions: 0,
            total_deletions: 0,
            accepted_lines: 0,
            overriden_lines: 0,
        },
    );

    working_log
        .write_initial_attributions(initial_attributions, prompts)
        .expect("write initial attributions should succeed");

    // NOW create both files - fileA gets INITIAL, fileB uses blame
    std::fs::write(repo.path().join("fileA.txt"), "line 1 in A\nline 2 in A\n")
        .expect("write file should succeed");
    std::fs::write(repo.path().join("fileB.txt"), "line 1 in B\nline 2 in B\n")
        .expect("write file should succeed");

    // Run checkpoint with mock_ai so fileB gets mock attribution
    repo.git_ai(&["checkpoint", "mock_ai"])
        .expect("checkpoint should succeed");

    // Commit
    repo.stage_all_and_commit("add both files")
        .expect("commit should succeed");

    // Check blame for fileA - should show INITIAL attributions (toolA)
    let blame_a = repo
        .git_ai(&["blame", "fileA.txt"])
        .expect("blame should succeed");

    let normalized_a = normalize_blame_output(&blame_a);
    assert_debug_snapshot!(normalized_a);

    // Check blame for fileB - should show mock (no INITIAL, so blame is used)
    let blame_b = repo
        .git_ai(&["blame", "fileB.txt"])
        .expect("blame should succeed");
    let normalized_b = normalize_blame_output(&blame_b);
    assert_debug_snapshot!(normalized_b);
}
