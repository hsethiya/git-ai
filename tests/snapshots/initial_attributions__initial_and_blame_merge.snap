---
source: tests/initial_attributions.rs
expression: normalized
---
"COMMIT_SHA (tool1 TIMESTAMP 1) line 1\nCOMMIT_SHA (tool1 TIMESTAMP 2) line 2\nCOMMIT_SHA (tool1 TIMESTAMP 3) line 3\nCOMMIT_SHA (mock_ai TIMESTAMP 4) line 4\nCOMMIT_SHA (tool2 TIMESTAMP 5) line 5\nCOMMIT_SHA (mock_ai TIMESTAMP 6) line 6\nCOMMIT_SHA (mock_ai TIMESTAMP 7) line 7\n"
